src/box/iproto.cc

```diff
@@ -3778,39 +3849,30 @@ iproto_override(uint32_t req_type, iproto_handler_t cb,
 		return -1;
 	}
 
-	if (cb == NULL) {
-		mh_int_t k = mh_i32ptr_find(tx_req_handlers, req_type, NULL);
-		if (k != mh_end(tx_req_handlers)) {
-			struct mh_i32ptr_node_t *node =
-				mh_i32ptr_node(tx_req_handlers, k);
-			struct iproto_req_handler *old =
-				(struct iproto_req_handler *)node->val;
-			iproto_cfg_override(req_type, false);
-			iproto_req_handler_delete(old);
-			mh_i32ptr_del(tx_req_handlers, k, NULL);
-		}
-		return 0;
-	}
+	struct iproto_req_handlers *handlers;
+	handlers = iproto_req_handlers_get(req_type);
+	bool old_is_set = iproto_req_handler_is_set(handlers);
 
-	struct iproto_req_handler *handler =
-		(struct iproto_req_handler *)xmalloc(sizeof(*handler));
-	*handler = (struct iproto_req_handler) {
-		/* .handler = */ cb,
-		/* .destroy = */ destroy,
-		/* .ctx = */ ctx,
-	};
-	struct mh_i32ptr_node_t node = {
-		/* .key = */ req_type,
-		/* .val = */ handler,
-	};
-	struct mh_i32ptr_node_t old;
-	struct mh_i32ptr_node_t *replaced = &old;
-	mh_i32ptr_put(tx_req_handlers, &node, &replaced, NULL);
-	if (replaced != NULL) {
-		handler = (struct iproto_req_handler *)replaced->val;
-		iproto_req_handler_delete(handler);
-		return 0;
+	if (handlers != NULL && handlers->c.destroy != NULL)
+		handlers->c.destroy(handlers->c.ctx);
+
+	if (cb != NULL) {
+		if (handlers == NULL)
+			handlers = iproto_req_handlers_new(req_type);
+		handlers->c.cb = cb;
+		handlers->c.destroy = destroy;
+		handlers->c.ctx = ctx;
+	} else if (handlers != NULL) {
+		handlers->c.cb = NULL;
+		handlers->c.destroy = NULL;
+		handlers->c.ctx = NULL;
 	}
-	iproto_cfg_override(req_type, true);
+
+	bool new_is_set = iproto_req_handler_is_set(handlers);
+	if (!new_is_set && handlers != NULL)
+		iproto_req_handlers_delete(handlers, req_type);
```

Comment:
Looks like a bug: you can delete a handler from TX only after notifying IPROTO. Otherwise there can be inflight requests that arrive at the override route only to see there is no request handler set (which is unexpected).

Reply:
Thanks, fixed.